Move to generic workload platform flow (#8)
ietf-wg-wimse/draft-ietf-wimse-workload-identity-bcp: draft-ietf-wimse-client-assertion-in-workload-environments.md

Workloads often require access to external resources to perform their tasks. For example, access to a database, a web server or another workload. These resources are protected by an authorization server and can only be accessed with an access token. The challenge for workloads is to get this access token issued.

Traditionally, workloads can be provisioned with client secrets credentials and use the client_credential flow to retrieve an access token. This model comes with a set of challenges that make it insecure and high-maintaince. Client secrets need to be manually provisioned and rotated. They can be stolen and used by attackers to impersonate the workload.

A solution to this problem is to not provision secret material to the workload itself but use the workload platform to attest for that workload. Many workload platforms offer a credential, in most cases a JWT token. Signed by a platform-internal authorization server, this credential attests the workload and its attributes. Based on {{RFC7521}} and its JWT profile {{RFC7523}}, this credential can then be used as a client assertion towards a different authorization server. {{fig-overview}} illustrates this interaction.

> ~~~

Author: Awesome, I feel like a Picasso now 🚀 😄

* A1) The workload directly access the protected resource with the credential issued in step 1.

In case the workload needs to access a resource outside of the cloud (protected by a different authorization server). This can also be the same cloud but different context (tenant, account).

* B1) The workload presents cloud-issued credential as an assertion towards the external authorization server using {{RFC7523}}. An access token is returned to the workload.

* B2) Using the access token from step B1, the workload is able to access the protected resource in the external authorization domain.

## Continues integration/deployment systems {#cicd}

Continous integration and deployment systems allow their pipelines/workflows to receive identity every time they run. Particularly in situations where build outputs need to be uploaded to resources protected by other authorization server, deployments need to be made, or more generally, protected resources to be accessed, {{RFC7523}} is used to federate the pipeline/workflow identity to an identity of the other authorization server.

> ~~~

Reviewer: ```suggestion
~~~ aasvg
```

 +---------+---------+-----+--------------------------------+
 |         |         |     |             Workload Platform  |
 |         |         |     |                                |
 |  +------+---------v-----+---+           +-------------+  |
 |  |                          |           | Credential  |  |
 |  |        Workload          +-----------> issued by   |  |
 |  |                          |  1) get   | Platform    |  |
 |  +--------------------------+           +-------------+  |
 +----------------------------------------------------------+
~~~
{: #fig-overview title="OAuth2 Assertion Flow in generic Workload Environment"}

The figure outlines the following steps which are applicable in any pattern.

* 1) retrieve credential issued by platform. The way this is acchieved and whether this is workload or platform initiated differs based on the platform.

* 2) present credential as an assertion towards an authorization server in an external authorization domain. This step uses the assertion_grant flow defined in {{RFC7521}} and, in case of JWT format, {{RFC7523}}.

* 3) On success, an access token is returned to the workload to access the protected resource.

> * 4) The aaccess token is used to access the protected resource in the external authorization domain.

Reviewer 2: Typo, aaccess -> access

> * 4) The aaccess token is used to access the protected resource in the external authorization domain.

Reviewer 3: per @edoardogiordano ...
```suggestion
* 4) The access token is used to access the protected resource in the external authorization domain.
```

---

> title: OAuth 2.0 Client Assertion in Workload Environments

Reviewer 4: I think "runtime environments" is a more common name.

Author: Let's discuss title in #6

     target: https://openid.net/specs/openid-connect-core-1_0.html
     date: 8 November 2014
  KubernetesServiceAccount:
     title: Kubernetes Service Account
     target: https://kubernetes.io/docs/concepts/security/service-accounts/
     date: 10 May 2024
  TokenReviewV1:
     title: Kubernetes Token Review API V1
     target: https://kubernetes.io/docs/reference/kubernetes-api/authentication-resources/token-review-v1/
     date: 28 August 2024
  TokenRequestV1:
     title: Kubernetes Token Request API V1
     target: https://kubernetes.io/docs/reference/kubernetes-api/authentication-resources/token-request-v1/
     date: 28 August 2024

--- abstract

The use of the OAuth 2.0 framework for container orchestration systems poses a challenge as managing secrets, such as client_id and client_secret, can be complex and error-prone. Instead of manual provisioning these credentials the industry has moved to a federation-based approach where credentials of the underlying workload platform are used as assertions towards an OAuth authorization server leveraging the Client Assertion Flow {{RFC7521}}, in particular {{RFC7523}}.

> This specifications describes a meta flow in {{overview}}, gives security recommendations in {{recommendations}} and outlines concrete patterns in {{patterns}}.

Reviewer 4: ```suggestion
This specification describes a meta flow in {{overview}}, gives security recommendations in {{recommendations}} and outlines concrete patterns in {{patterns}}.
```

But also you don't normally dive down into sections in the abstract. This works better in the Intro.

Reviewer 3: that guy makes a good point

Workloads often require access to external resources to perform their tasks. For example, access to a database, a web server or another workload. These resources are protected by an authorization server and can only be accessed with an access token. The challenge for workloads is to get this access token issued.

Traditionally, workloads were provisioned with client credentials and use the corresponding client credential flow (Section 1.3.4 {{RFC6749}}) to retrieve an access token. This model comes with a set of challenges that make it insecure and high-maintenance. Secret materials need to be provisioned and rotated, which often happens manually. It also can be stolen and used by attackers to impersonate the workload.

A solution to this problem is to not provision secret material to the workload and use the platform the workload runs on to attest for it. Many workload platforms offer a credential, in most cases a JWT token. Signed by a platform-internal authorization server, the credential attests the workload and its attributes. Based on {{RFC7521}} and its JWT profile {{RFC7523}}, this credential can then be used as a client assertion towards a different authorization server.

> 

Reviewer 4: I suggest adding a paragraph saying something like:

> This document is published by the WIMSE working group. However it refers to existing industry practices which may or may not be in line with the (currently work in progress) WIMSE architecture and the protocols derived from it. Practitioners are advised to keep that in mind as they devise their adoption strategy.

Author: Created #9

~~~ aasvg
 +----------------------------------------------------------+
 |                           External Authorization Domain  |
 |                                                          |
 | +-------------------------+     +--------------------+   |
 | |                         |     |                    |   |
 | |   Authorization Server  |     | Protected Resource |   |
 | |                         |     |                    |   |
 | +-------^---------+-------+     +----------^---------+   |
 +---------+---------+------------------------+-------------+
           |         |                        |
           |   3) access token           4) access
           |         |                        |
2) present assertion |                        |
           |         |     +------------------+
           |         |     |
 +---------+---------+-----+--------------------------------+
 |         |         |     |             Workload Platform  |
 |         |         |     |                                |
 |  +------+---------v-----+---+           +-------------+  |
 |  |                          |           | Credential  |  |

>  |  |        Workload          +-----------> issued by   |  |

Reviewer 5: ```suggestion
 |  |        Workload          +<---------> issued by   |  |
```

Reviewer: @hannestschofenig this edit misaligns the diagram, was the intent to just make it a double-headed arrow? If so that syntax would be as follows:

```suggestion
 |  |        Workload          <-----------> issued by   |  |
```

Author: Latest commit makes the arrow going both ways.

 |  |                          |  1) get   | Platform    |  |
 |  +--------------------------+           +-------------+  |
 +----------------------------------------------------------+
~~~
{: #fig-overview title="OAuth2 Assertion Flow in generic Workload Environment"}

The figure outlines the following steps which are applicable in any pattern.

* 1) retrieve credential issued by platform. The way this is achieved and whether this is workload or platform initiated differs based on the platform.

* 2) present credential as an assertion towards an authorization server in an external authorization domain. This step uses the assertion_grant flow defined in {{RFC7521}} and, in case of JWT format, {{RFC7523}}.

* 3) On success, an access token is returned to the workload to access the protected resource.

> * 4) The access token is used to access the protected resource in the external authorization domain.

Reviewer 4: This terminology is very confusing to first-timers. Please say, at least once, that when we discuss "accessing a resource" we also include "sending a message" or "making a REST call".

Accessing different protected resources may require steps 2) to 4) again with different scope parameters. Accessing a protected resource in an entirely different authorization domain often requires the entire flow to be followed again, to retrieve a new platform-issued credential with an audience for the external authorization server. This, however, differs based on the platform and implementation.

## Credential format

For the scope of this document we focus on JSON Web Token credential formats. Other formats such as X.509 certificates are possible but not as widely seen as JSON Web Tokens.

The claims in the present assertion vary greatly based on use case and actual platform, but a minimum set of claims are seen across all of them. {{RFC7523}} describes them in detail and according to it, all MUST be present.

~~~json
{
  "iss": "https://example.org",
  "sub": "my-workload",

>   "aud": "custom-audience",

Reviewer 4: target-audience?

Reviewer 5: target audience is better than custom audience.

      "name": "127.0.0.1",
      "uid": "58456cb0-dd00-45ed-b797-5578fdceaced"
    },
    "pod": {
      "name": "my-workload-69cbfb9798-jv9gn",
      "uid": "778a530c-b3f4-47c0-9cd5-ab018fb64f33"
    },
    "serviceaccount": {
      "name": "my-workload",
      "uid": "a087d5a0-e1dd-43ec-93ac-f13d89cd13af"
    },
    "warnafter": 1700081020
  },
  "nbf": 1700077413,
  "sub": "system:serviceaccount:my-namespace:my-workload"
}
~~~
{: #fig-kubernetes-token title="Example Kubernetes Service Account Token claims"}

## Secure Production Identity Framework For Everyone (SPIFFE) {#spiffe}

Secure Production Identity Framework For Everyone, also known as SPIFFE, is a cloud native compute foundation (CNCF) adopted project which defines an API defined called "Workload API" to delivery machine identity to workloads. Workloads can retrieve either X509 based or JWT credentials without the need to authenticate making it very easy to use. How workloads authenticate on the API is not part of the specification. It is common to use platform metadata from the operating system and the workload platform for authentication on the Workload API.

> For the scope of this document, the JWT formatted credential is the most relevant one. SPIFFE referres to it as "JWT-SVID" (JWT - Single Verifyable Identity Document).

Reviewer 6: ```suggestion
For the scope of this document, the JWT formatted credential is the most relevant one. SPIFFE refers to it as "JWT-SVID" (JWT - SPIFFE Verifiable Identity Document).
```
SVID is short for SPIFFE Verifiable Identity Document, see https://spiffe.io/docs/latest/spiffe-about/spiffe-concepts/#spiffe-verifiable-identity-document-svid

Author: Thank you!

│ │                         │     │                    │   │
│ │   Authorization Server  │     │  Resource Server   │   │
│ │                         │     │                    │   │
│ └────────────┬────────────┘     └────────────▲───────┘   │
│              │                               │           │
└──────────────┼───────────────────────────────┼───────────┘
               │                               │
     2) present assertion                 3) access
               │                               │
               │     ┌─────────────────────────┘
               │     │
┌──────────────┼─────┼─────────────────────────────────────┐
│              │     │                  Workload Platform  │
│              │     │                                     │
│  ┌───────────┴─────┴───┐           ┌──────────────────┐  │
│  │                     │           │ Credential       │  │
│  │  Workload           ├───────────► issued by        │  │
│  │                     │  1) get   │ Platform         │  │
│  └─────────────────────┘           └──────────────────┘  │
│                                                          │
│                                                          │

> └──────────────────────────────────────────────────────────┘

Reviewer 7: I Like the new diagram, Its much more agnostic and to k8s, but still shows the approach clearly.

The terms 'workload' and 'container' are used interchangeably.

# Patterns {#patterns}

## Kubernetes {#kubernetes}

> Kubernetes notion of identity is mainly the concept of "service accounts" which are either created explictly or automatically during deployment time. While the mean purpose of these identities is to authenticate access to the Kubernetes Control Plane it is more and more used to authenticate workloads between each other too. The credential format used to authenticate service account is JWT that are signed by the Kubernetes Control Plane.

Reviewer 7: >it is more and more used to authenticate workloads between each other too

**and authenticate agains other domains such as cloud accounts.**

Author: I rewrote the section a bit, Kubernetes mentions the service account use cases here https://kubernetes.io/docs/concepts/security/service-accounts/#use-cases.

Before, I wrote that primary use case is to authenticate Kubernetes APIs, but I guess that's quite a old statement.

Also added references to Kubernetes docs.

To programatically use service account tokens workloads can

* use the TokenRequest API of the control plane

* have the token projected into the file system of the workload. This is commonly referred to as "projected service accout token".

Both options allow workloads to

* specify a custom audience. Possible audiences can be restricted based on control plane policy.

> * specify a custom lifetime. (TODO, is policy possible?)

Reviewer 7: https://kubernetes.io/docs/concepts/storage/projected-volumes/#serviceaccounttoken
Yes - one can specify a custom lifetime when projecting the service account token into a pod's filesystem.

Author: Thank you!
